Support Reading HTTP Response When Request Isn't Fully Sent

#### Environment details

- OS: Ubuntu
- Java version: openjdk-1.8.0
- google-http-java-client version: 1.23

#### Steps to reproduce

1. Use the HTTP Client to send a large post request that aborts early (e.g. permission denied to any Google API)

#### Stacktrace

```
org.dcm4che3.net.service.DicomServiceException: java.io.IOException: Error writing request body to server( status = 272 )
        at com.google.cloud.healthcare.imaging.dicomadapter.CStoreService.store(CStoreService.java:81)
        at org.dcm4che3.net.service.BasicCStoreSCP.onDimseRQ(BasicCStoreSCP.java:72)
        at org.dcm4che3.net.service.DicomServiceRegistry.onDimseRQ(DicomServiceRegistry.java:86)
        at org.dcm4che3.net.ApplicationEntity.onDimseRQ(ApplicationEntity.java:485)
        at org.dcm4che3.net.Association.onDimseRQ(Association.java:650)
        at org.dcm4che3.net.PDUDecoder.decodeDIMSE(PDUDecoder.java:459)
        at org.dcm4che3.net.Association.handlePDataTF(Association.java:634)
        at org.dcm4che3.net.State$4.onPDataTF(State.java:103)
        at org.dcm4che3.net.Association.onPDataTF(Association.java:630)
        at org.dcm4che3.net.PDUDecoder.nextPDU(PDUDecoder.java:177)
        at org.dcm4che3.net.Association$2.run(Association.java:478)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Error writing request body to server
        at sun.net.www.protocol.http.HttpURLConnection$StreamingOutputStream.checkError(HttpURLConnection.java:3587)
        at sun.net.www.protocol.http.HttpURLConnection$StreamingOutputStream.write(HttpURLConnection.java:3570)
        at com.google.api.client.util.ByteStreams.copy(ByteStreams.java:55)
        at com.google.api.client.util.IOUtils.copy(IOUtils.java:94)
        at com.google.api.client.http.AbstractInputStreamContent.writeTo(AbstractInputStreamContent.java:72)
        at com.google.api.client.http.MultipartContent.writeTo(MultipartContent.java:107)
        at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:80)
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
        at com.google.cloud.healthcare.imaging.dicomadapter.CStoreService.store(CStoreService.java:78)
        ... 13 more
```

#### External references such as API reference guides used

#### Any additional information below

This was identified by using the following tool: https://github.com/GoogleCloudPlatform/healthcare/tree/master/imaging/dicom_adapter#import-adapter which uses this HTTP Client underneath

I've played around with the library a bit but can't seem to find any way for it to return meaningful errors in this context. If I send a regular curl request it handles this just find and return the actual HTTP Response. 

Making sure to follow these steps will guarantee the quickest resolution possible.

Thanks!

--- a/google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpRequest.java
+++ b/google-http-client/src/main/java/com/google/api/client/http/javanet/NetHttpRequest.java
@@ -117,6 +117,12 @@ final class NetHttpRequest extends LowLevelHttpRequest {
           writeContentToOutputStream(outputWriter, out);
 
           threw = false;
+        } catch (IOException e) {
+          // If we've gotten a response back, continue on and try to parse the response. Otherwise,
+          // re-throw the IOException
+          if (!hasReponse(connection)) {
+            throw e;
+          }
         } finally {
           try {
             out.close();
@@ -147,6 +153,15 @@ final class NetHttpRequest extends LowLevelHttpRequest {
       if (!successfulConnection) {
         connection.disconnect();
       }
+    }
+  }
+
+  private boolean hasReponse(HttpURLConnection connection) {
+    try {
+      return connection.getResponseCode() > 0;
+    } catch (IOException e) {
+      // There's some exception trying to parse the response
+      return false;
     }
   }
 

--- a/google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpRequestTest.java
+++ b/google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpRequestTest.java
@@ -165,4 +165,36 @@ public class NetHttpRequestTest {
       assertEquals("Error writing request body to server", e.getMessage());
     }
   }
+
+  @Test
+  public void testErrorOnClose() throws Exception {
+    MockHttpURLConnection connection = new MockHttpURLConnection(new URL(HttpTesting.SIMPLE_URL)) {
+      @Override
+      public OutputStream getOutputStream() throws IOException {
+        return new OutputStream() {
+          @Override
+          public void write(int b) throws IOException {
+            return;
+          }
+
+          @Override
+          public void close() throws IOException {
+            throw new IOException("Error during close");
+          }
+        };
+      }
+    };
+    connection.setRequestMethod("POST");
+    NetHttpRequest request = new NetHttpRequest(connection);
+    InputStream is = NetHttpRequestTest.class.getClassLoader().getResourceAsStream("file.txt");
+    HttpContent content = new InputStreamContent("text/plain", is);
+    request.setStreamingContent(content);
+
+    try {
+      request.execute();
+      fail("Expected to throw an IOException");
+    } catch (IOException e) {
+      assertEquals("Error during close", e.getMessage());
+    }
+  }
 }

--- a/google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpRequestTest.java
+++ b/google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpRequestTest.java
@@ -134,4 +134,35 @@ public class NetHttpRequestTest {
     }
   }
 
+  @Test
+  public void testInterruptedWriteErrorOnResponse() throws Exception {
+    MockHttpURLConnection connection = new MockHttpURLConnection(new URL(HttpTesting.SIMPLE_URL)) {
+      @Override
+      public OutputStream getOutputStream() throws IOException {
+        return new OutputStream() {
+          @Override
+          public void write(int b) throws IOException {
+            throw new IOException("Error writing request body to server");
+          }
+        };
+      }
+
+      @Override
+      public int getResponseCode() throws IOException {
+        throw new IOException("Error parsing response code");
+      }
+    };
+    connection.setRequestMethod("POST");
+    NetHttpRequest request = new NetHttpRequest(connection);
+    InputStream is = NetHttpRequestTest.class.getClassLoader().getResourceAsStream("file.txt");
+    HttpContent content = new InputStreamContent("text/plain", is);
+    request.setStreamingContent(content);
+
+    try {
+      request.execute();
+      fail("Expected to throw an IOException");
+    } catch (IOException e) {
+      assertEquals("Error writing request body to server", e.getMessage());
+    }
+  }
 }

--- a/google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpRequestTest.java
+++ b/google-http-client/src/test/java/com/google/api/client/http/javanet/NetHttpRequestTest.java
@@ -2,6 +2,7 @@ package com.google.api.client.http.javanet;
 
 import com.google.api.client.http.HttpContent;
 import com.google.api.client.http.InputStreamContent;
+import com.google.api.client.http.LowLevelHttpResponse;
 import com.google.api.client.http.javanet.NetHttpRequest.OutputWriter;
 import com.google.api.client.testing.http.HttpTesting;
 import com.google.api.client.testing.http.javanet.MockHttpURLConnection;
@@ -80,6 +81,57 @@ public class NetHttpRequestTest {
     request.setStreamingContent(content);
     request.setWriteTimeout(timeout);
     request.execute(new SleepingOutputWriter(5000L));
+  }
+
+  @Test
+  public void testInterruptedWriteWithResponse() throws Exception {
+    MockHttpURLConnection connection = new MockHttpURLConnection(new URL(HttpTesting.SIMPLE_URL)) {
+      @Override
+      public OutputStream getOutputStream() throws IOException {
+        return new OutputStream() {
+          @Override
+          public void write(int b) throws IOException {
+            throw new IOException("Error writing request body to server");
+          }
+        };
+      }
+    };
+    connection.setResponseCode(401);
+    connection.setRequestMethod("POST");
+    NetHttpRequest request = new NetHttpRequest(connection);
+    InputStream is = NetHttpRequestTest.class.getClassLoader().getResourceAsStream("file.txt");
+    HttpContent content = new InputStreamContent("text/plain", is);
+    request.setStreamingContent(content);
+
+    LowLevelHttpResponse response = request.execute();
+    assertEquals(401, response.getStatusCode());
+  }
+
+  @Test
+  public void testInterruptedWriteWithoutResponse() throws Exception {
+    MockHttpURLConnection connection = new MockHttpURLConnection(new URL(HttpTesting.SIMPLE_URL)) {
+      @Override
+      public OutputStream getOutputStream() throws IOException {
+        return new OutputStream() {
+          @Override
+          public void write(int b) throws IOException {
+            throw new IOException("Error writing request body to server");
+          }
+        };
+      }
+    };
+    connection.setRequestMethod("POST");
+    NetHttpRequest request = new NetHttpRequest(connection);
+    InputStream is = NetHttpRequestTest.class.getClassLoader().getResourceAsStream("file.txt");
+    HttpContent content = new InputStreamContent("text/plain", is);
+    request.setStreamingContent(content);
+
+    try {
+      request.execute();
+      fail("Expected to throw an IOException");
+    } catch (IOException e) {
+      assertEquals("Error writing request body to server", e.getMessage());
+    }
   }
 
 }

